// larasocket-js/dist/util/link-message.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.LinkMessage = void 0;
var message_1 = require("./message");
/**
 * LinkMessage
 */
var LinkMessage = /** @class */ (function (_super) {
    __extends(LinkMessage, _super);
    /**
     * Create a new class instance.
     */
    function LinkMessage(token, uuid) {
        var _this = _super.call(this) || this;
        _this.token = token;
        _this.uuid = uuid;
        return _this;
    }
    /**
     *
     */
    LinkMessage.prototype.toNetworkJson = function () {
        return {
            action: 'link',
            token: this.token,
            uuid: this.uuid,
        };
    };
    return LinkMessage;
}(message_1.Message));
exports.LinkMessage = LinkMessage;
