// larasocket-js/dist/util/outgoing-message.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.OutgoingMessage = exports.OutgoingMessageType = void 0;
var message_1 = require("./message");
var OutgoingMessageType;
(function (OutgoingMessageType) {
    OutgoingMessageType["LINK_CONNECTION"] = "link";
    OutgoingMessageType["SUBSCRIBE"] = "subscribe";
    OutgoingMessageType["UNSUBSCRIBE"] = "unsubscribe";
    OutgoingMessageType["WHISPER"] = "whisper";
    OutgoingMessageType["BROADCAST"] = "broadcast";
})(OutgoingMessageType = exports.OutgoingMessageType || (exports.OutgoingMessageType = {}));
/**
 * OutgoingMessage
 */
var OutgoingMessage = /** @class */ (function (_super) {
    __extends(OutgoingMessage, _super);
    /**
     * Create a new class instance.
     */
    function OutgoingMessage(token, action) {
        var _this = _super.call(this) || this;
        _this.token = token;
        _this.action = action;
        return _this;
        // this.connectionId = connectionId;
    }
    /**
     *
     */
    OutgoingMessage.prototype.toNetworkJson = function () {
        var _a;
        return {
            event: this.event,
            token: this.token,
            connection_id: this.connectionId,
            action: this.action,
            payload: this.payload,
            channel: (_a = this.channel) === null || _a === void 0 ? void 0 : _a.name,
        };
    };
    return OutgoingMessage;
}(message_1.Message));
exports.OutgoingMessage = OutgoingMessage;
