// larasocket-js/dist/connector/connector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Connector = void 0;
var Connector = /** @class */ (function () {
    /**
     * Create a new class instance.
     */
    function Connector(options) {
        /**
         * Default connector options.
         */
        this._defaultOptions = {
            auth: {
                headers: {},
            },
            authEndpoint: '/broadcasting/auth',
            broadcaster: 'pusher',
            csrfToken: null,
            host: null,
            key: null,
            namespace: 'App.Events',
        };
        this.setOptions(options);
        this.connect();
    }
    /**
     * Merge the custom options with the defaults.
     */
    Connector.prototype.setOptions = function (options) {
        this.options = Object.assign(this._defaultOptions, options);
        if (this.csrfToken()) {
            this.options.auth.headers['X-CSRF-TOKEN'] = this.csrfToken();
        }
        return options;
    };
    /**
     * Extract the CSRF token from the page.
     */
    Connector.prototype.csrfToken = function () {
        var selector;
        // @ts-ignore
        if (typeof window !== 'undefined' && window.Laravel && window.Laravel.csrfToken) {
            // @ts-ignore
            return window.Laravel.csrfToken;
        }
        else if (this.options.csrfToken) {
            return this.options.csrfToken;
        }
        else if (typeof document !== 'undefined' && typeof document.querySelector === 'function') {
            selector = document.querySelector('meta[name="csrf-token"]');
            if (selector) {
                return selector.getAttribute('content');
            }
        }
        return null;
    };
    return Connector;
}());
exports.Connector = Connector;
